test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -1051,32 +1079,96 @@ class test_ucp_sockaddr_protocols : public test_ucp_sockaddr {
+                                            scomplete_err_handling_cb);
             } else {
                 sreq = ucp_tag_send_nb(sender().ep(), &send_buf[0], size,
-                                       ucp_dt_make_contig(1), 0, scomplete_cb);
+                                       ucp_dt_make_contig(1), 0,
+                                       !err_handling ? scomplete_cb :
+                                       scomplete_err_handling_cb);
             }
 
             if (!is_exp) {
-                short_progress_loop();
-                rreq = ucp_tag_recv_nb(receiver().worker(), &recv_buf[0], size,
-                                       ucp_dt_make_contig(1), 0, 0,
-                                       rtag_complete_cb);
+                ucp_tag_recv_info_t recv_info = {};
+                ucp_tag_message_h message;
+                do {
+                    short_progress_loop();
+                    message = ucp_tag_probe_nb(receiver().worker(),
+                                               0, 0, 1, &recv_info);
+                } while (message == NULL);
+
+                EXPECT_EQ(size, recv_info.length);
+                EXPECT_EQ(0,    recv_info.sender_tag);
+
+                if (err_handling) {
+                    if (recv_stop != NULL) {
+                        recv_stop(arg);
+                    }
+
+                    /* If send request is NULL, it means that send operation
+                     * completed immediately */
+                    if (sreq != NULL) {
+                        /* TODO: remove memory deregistration, when UCP
+                         * requests tracking is added */
+                        ucp_request_t *req = static_cast<ucp_request_t*>
+                                             (sreq) - 1;
+                        EXPECT_EQ(sender().ucph(),
+                                  req->send.ep->worker->context);
+                        ucp_request_memory_dereg(sender().ucph(),
+                                                 req->send.datatype,
+                                                 &req->send.state.dt, req);
+                    }
+
+                    if (recv_stop != NULL) {
+                        sender().disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
+                    } else {
+                        send_stop(arg);
+                        receiver().disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
+                    }
+                }
+
+                rreq = ucp_tag_msg_recv_nb(receiver().worker(), &recv_buf[0],
+                                           size, ucp_dt_make_contig(1), message,
+                                           !err_handling ? rtag_complete_cb :
+                                           rtag_complete_err_handling_cb);
             }
 
-            request_wait(sreq);
-            request_wait(rreq);
+            if (!err_handling) {
+                request_wait(sreq);
+                request_wait(rreq);
```

Comment:
make std::vector of the requests we need to wait for

Reply:
done